rethinkdb/handshake.py

```diff
@@ -35,306 +35,328 @@
+                version=self._protocol_version,
+                min=min_protocol_version,
+                max=max_protocol_version
+            ))
+
+        self._next_state()
+        return ''
+
+    def _prepare_auth_request(self, response):
+        """
+        Put tohether the authentication request based on the response of the database.
+
+        :param response: Response from the database
+        :raises: ReqlDriverError | ReqlAuthError
+        :return: An empty string
+        """
+
+        json_response = self._decode_json_response(response, with_utf8=True)
+        first_client_message = json_response['authentication'].encode('ascii')
+
+        authentication = dict(x.split(b'=', 1) for x in first_client_message.split(b','))
+
+        r = authentication[b'r']
+        if not r.startswith(self._r):
+            raise ReqlAuthError('Invalid nonce from server', self._host, self._port)
+
+        salted_password = self._pbkdf2_hmac(
+            'sha256',
+            self._password,
+            base64.standard_b64decode(authentication[b's']),
+            int(authentication[b'i'])
+        )
+
+        message_without_proof = b'c=biws,r={r}'.format(r=r)
+        auth_message = b','.join((
+            self._first_client_message,
+            first_client_message,
+            message_without_proof
+        ))
+
+        self._server_signature = hmac.new(
+            hmac.new(salted_password, b'Server Key', hashlib.sha256).digest(),
+            auth_message,
+            hashlib.sha256
+        ).digest()
+
+        client_key = hmac.new(salted_password, b'Client Key', hashlib.sha256).digest()
+        client_signature = hmac.new(hashlib.sha256(client_key).digest(), auth_message, hashlib.sha256).digest()
+        client_proof = struct.pack('32B', *(l ^ r for l, r in zip(
+            struct.unpack('32B', client_key),
+            struct.unpack('32B', client_signature)
+        )))
+
+        authentication_request = bytes('{auth_request}\0'.format(auth_request=self._json_encoder.encode({
+            'authentication': bytes('{message_without_proof},p={proof}'.format(
+                message_without_proof=message_without_proof,
+                proof=base64.standard_b64encode(client_proof)
+            )).decode('ascii')
+        }).encode('utf-8')))
+
+        self._next_state()
+        return authentication_request
+
+    def _read_auth_response(self, response):
+        """
+        Read the authentication request's response sent by the database.
+
+        :param response: Response from the database
+        :raises: ReqlDriverError | ReqlAuthError
+        :return: An empty string
+        """
+
+        json = self._json_decoder.decode(response.decode('utf-8'))
+        v = None
+        try:
+            if json['success'] is False:
+                if 10 <= json['error_code'] <= 20:
+                    raise ReqlAuthError(json['error'], self._host, self._port)
+                else:
```

Comment:
No need for else here.

Reply:
Yes, this function is not refactored yet. This is the only missing piece right now (and the unit tests of course).